libgalois/include/galois/Reduction.h

```diff
@@ -20,546 +20,217 @@
 #ifndef GALOIS_REDUCTION_H
 #define GALOIS_REDUCTION_H
 
-#include "galois/gstl.h"
 #include "galois/substrate/PerThreadStorage.h"
 
+#include <functional>
 #include <limits>
 
 namespace galois {
 
 /**
- * GSimpleReducible stores per thread values of a variable of type T, suitable
- * for small, cheap to copy, plain T types, where T is not a container, is small
- * and does not allocate memory
+ * A Reducible stores per-thread values of a variable of type T and merges
+ * multiple values into one.
  *
- * The final value is obtained by performing a reduction on per thread values
- * using the provided binary functor BinFunc. BinFunc conforms to STL semantics,
- * such as std::plus<T>, with signature:
- *  T operator()(const T& lhs, const T& rhs)
+ * The reduced value is obtained by merging per thread values using the binary
+ * functor MergeFunc. MergeFunc takes two values of type T and produces the
+ * resulting merged value:
  *
- * The identity value provided to the constructor is the identity element for
- * the binary functor. i.e. f(x, identity) == x
+ *   T operator()(T lhs, T rhs)
+ *
+ * If T is expensive to copy, a moving merge function is more appropriate:
+ *
+ *   T& operator()(T& lhs, T&& rhs)
+ *
+ * IdFunc returns the identity element, which is used to initialize and reset
+ * the per thread values.
+ *
+ * The MergeFunc and IdFunc should have following property:
```

Comment:
They also need to be copy constructable. Lambdas happen to get this for free.

Reply:
Updated function doc.